### DISCUSSION TITLE
Registering Package Advice

### DISCUSSION THREAD

**Discussion by nas2011** [toitlang/toit#608] [Drivers]
Hello. I have written a minimal driver for the ADXL345 accelerometer and intend to add more features to it as I go. I am 100% a practitioner NOT a developer so my number one deliverable is a working end product. I try to write code that follows best practices but sometimes it's just a "get it done" proposition. Should I bother to try and register it as a package if I'm not checking all the convention and best practice boxes?

In the course of my work I feel like I build some useful code (even if it is niche) but I never share or publish anything because my criteria is more along the lines of can it effectively measure motor vibration rather than is it clean code. Just not sure if it is worth putting out there. I'd appreciate any feedback anyone has.

**Comment by floitsch** [comment]
Sorry for the late reply (holidays and moving into a new office)...

I think publishing packages is fine even if they don't have great quality.
My suggestion is to use version numbers 0.X.Y for packages that aren't yet production ready. This means that 1.0.0 doesn't mean "feature-complete" (as is often the case for semver packages), but rather "ready to be used in production".

If you want to, I would also be happy to do a review on your packages to make sure they are consistent (naming, style, ...).

**Comment by nas2011** [comment]
Decided to just push this out there and use that as the motivating factor to work on it more. I really like toit and want to do what I can to encourage/help others use it.

Keep getting this error though when I try to publish:
![image](https://user-images.githubusercontent.com/101505861/184506772-a3efd1aa-c80b-4592-ae08-73e730eb046f.png)


I have a package.yaml file so I'm not sure where I'm going wrong. Repo is here: https://github.com/nas2011/ADXL345.toit

I'm sure it's simple, but could someone point out what I'm doing wrong? thanks in advance for the help.

**Reply by floitsch** [reply]
The reason the message says that the `package.yaml` is missing, is that the registry fetches the sources corresponding to the tag you gave it. Since the code tagged with `v1.0.0` only has a `Package.yaml` file, the registry server complains.
You can just delete your v1.0.0 tag, and reapply it on top of master to make things work.
Before you do that, though:

Note: the title of your README currently says "ADXL345.toit". This would give the package the name "adxl345_toit". I would change the title to "ADXL345" and/or add a "name: adxl345" line to your package.yaml.

Also, the privacy signal in Toit is a trailing "_". All the registers (`_REG_DEVID`, ...) should be changed to have the `_` at the end: `REG_DEVID_`.

Since you already have an example in your code, you might as well move it to `examples/main.toit`. Generally, we don't use any `default_i2c`, as the ESP32 really allows to connect devices to pretty much any pin, but feel free to keep that one. See the toit-lsm303d for a typical example: https://github.com/toitware/toit-lsm303d/blob/main/examples/main.toit
It also shows how we often have an `I2C_ADDRESS` (and a `I2C_ADDRESS_ALT`) in the driver class. Feel free to ping me, if you want me to do a "real" review.
